### DISCUSSION TITLE
Ununderstandible issue with multithreading

### DISCUSSION THREAD

**Discussion by stserakhau** [devopvoid/webrtc-java#102] [General]
Strange behaviour of the code
```
vvvvvvvvvvvvvvvvv CODE 1 vvvvvvvvvvvvvvvv
RTCDataChannelInit init = new RTCDataChannelInit();
init.maxPacketLifeTime = 3000;
init.priority = RTCPriorityType.HIGH;
init.ordered = true;
String channelName = "media/" + taskUuid;
RTCDataChannel channel = rtcClient.peerConnection.createDataChannel(channelName, init);
MediaChannelObserver observer = new MediaChannelObserver(mediaCaptureCallbackSupport, channel);
channel.registerObserver(observer);

class MediaChannelObserver  {
....
        public void onStateChange() {
            log.debug("Media channel state changed to: {}", mediaChannel.getState());
            switch (mediaChannel.getState()) {
                case CLOSING -> close();                            <<<<< line 1
                case OPEN -> open();                                  <<<<< line 2
            }
        }
....
}
^^^^^^^^^^^^^^^^^^ CODE 1 ^^^^^^^^^^^^^^^^^^^^^
```

when I put execution of this code inside of ThreadPoolExecutor
SYSTEM_EXECUTOR.execute(()->{
  call CODE 1
})

Observer never executed onStateChange on OPEN event
CLOSE / CLOSING executed when second side executes CLOSE method on DataChannel.

workaround - replace OPEN event via sending message like 'OPEN' from second side and do initilization
CLOSE - the same via sending message or by timout

**Comment by devopvoid** [comment]
> Observer never executed onStateChange on OPEN event

It works with all events. Please make sure you create the data channel and register the observer prior to creating an offer or answer. Also run all the peer connection code in one the thread, e.g. your SYSTEM_EXECUTOR.

**Reply by stserakhau** [reply]
Hi, if I do the call in the same thread which received request from command datachannel all works fine, but when I put execution of the command which create new one data channel to another process I see some strange behaviour.

**Reply by devopvoid** [reply]
> when I put execution of the command which create new one data channel to another process

You are advised not to do so. Best practice is to create WebRTC related objects in a single thread. The observer callbacks are executed from native threads.

What you can do is to split the code you want to execute in other process.
